diffrax/_term.py

```diff
@@ -776,3 +785,180 @@ def _to_vjp(_y, _diff_args, _diff_term):
+    and velocity, $w$ is a Brownian motion in $\mathbb{R}^d$,
+    $f: \mathbb{R}^d \rightarrow \mathbb{R}$ is a potential function, and
+    $\gamma , u \in \mathbb{R}^{d \times d}$ are diagonal matrices governing
+    the friction and the inertia of the system.
+    """
+
+    gamma: PyTree[ArrayLike]
+    u: PyTree[ArrayLike]
+    control: AbstractBrownianPath
+
+    def vf(
+        self, t: RealScalarLike, y: UnderdampedLangevinTuple, args: Args
+    ) -> UnderdampedLangevinX:
+        x, v = y
+        # gamma, u and v can all have different pytree structures, we only know that
+        # gamma and u are prefixes of v
+        gamma = _broadcast_pytree(self.gamma, v)
+        u = _broadcast_pytree(self.u, v)
+
+        def _fun(_gamma, _u):
+            return jnp.sqrt(2 * _gamma * _u)
+
+        vf_v = jtu.tree_map(_fun, gamma, u)
+        return vf_v
+
+    def contr(
+        self, t0: RealScalarLike, t1: RealScalarLike, **kwargs
+    ) -> Union[UnderdampedLangevinX, AbstractBrownianIncrement]:
+        return self.control.evaluate(t0, t1, **kwargs)
+
+    def prod(
+        self, vf: UnderdampedLangevinX, control: UnderdampedLangevinX
+    ) -> UnderdampedLangevinTuple:
+        # The vf is only for the velocity component. The position component is
+        # unaffected by the diffusion.
+        dw = control
+        v_out = jtu.tree_map(operator.mul, vf, dw)
+        x_out = jtu.tree_map(jnp.zeros_like, v_out)
+        return x_out, v_out
+
+
+class UnderdampedLangevinDriftTerm(AbstractTerm):
+    r"""Represents the drift term in the Underdamped Langevin Diffusion (ULD).
+    The ULD SDE takes the form:
+
+    \begin{align*}
+        \mathrm{d} x(t) &= v(t) \, \mathrm{d}t \\
+        \mathrm{d} v(t) &= - \gamma \, v(t) \, \mathrm{d}t - u \,
+        \nabla \! f( x(t) ) \, \mathrm{d}t + \sqrt{2 \gamma u} \, \mathrm{d} w(t),
+    \end{align*}
+
+    where $x(t), v(t) \in \mathbb{R}^d$ represent the position
+    and velocity, $w$ is a Brownian motion in $\mathbb{R}^d$,
+    $f: \mathbb{R}^d \rightarrow \mathbb{R}$ is a potential function, and
+    $\gamma , u \in \mathbb{R}^{d \times d}$ are diagonal matrices governing
+    the friction and the inertia of the system.
+    """
+
+    gamma: PyTree[ArrayLike]
+    u: PyTree[ArrayLike]
+    grad_f: Callable[[UnderdampedLangevinX], UnderdampedLangevinX]
+
+    def vf(
+        self, t: RealScalarLike, y: UnderdampedLangevinTuple, args: Args
+    ) -> UnderdampedLangevinTuple:
+        x, v = y
+        # gamma, u and v can all have different pytree structures, we only know that
+        # gamma and u are prefixes of v (which is the same as x)
+        gamma = _broadcast_pytree(self.gamma, v)
+        u = _broadcast_pytree(self.u, v)
+
+        f_x = self.grad_f(x)
+        vf_x = v
+
+        def fun(_gamma, _u, _v, _f_x):
+            return -_gamma * _v - _u * _f_x
+
+        vf_v = jtu.tree_map(fun, gamma, u, v, f_x)
+        vf_y = (vf_x, vf_v)
+        return vf_y
```

Comment:
Because we went with the first option regarding when to broadcast `gamma` and `u` to the shape of `x` (as discussed [here](https://github.com/patrick-kidger/diffrax/pull/453#discussion_r1724906904)), we must perform `_broadcast_pytree(gamma)` (and `u`) each time we call `UnderdampedLangevin____Term.vf`. This isn't used in the ULD-specific solvers, but generic solvers call this all the time. I assume XLA optimises this, but do you think this would negatively affect compile time?

Maybe I could do the broadcast as part of the multiplication, but that might make the code a bit uglier and I'm not sure if it would significantly improve the situation.

What's your read on it?

Reply:
I think this looks good to me! I think these operations should all be fairly negligible for compile time.

I'm really happy with this change, it looks a lot cleaner to me.

Reply:
Great, that's what I was hoping for :)